### DISCUSSION TITLE
Porting Jainja JVM

### DISCUSSION THREAD

**Discussion by glegris** [vvaltchev/tilck#226] [Show and tell]
Hello,

I am the author of the [Jainja JVM](https://jainja.thenesis.org/). I have tried to make it work on Tilck and have obtained preliminary results.

Things that seem to work:
- Basic tests with java.lang
- Preemptive multithreading

Things not yet tested: java.io, java.nio, and low-level graphics

Current difficulties: I can't load classes from a jar file, so there may be problems with fseek

![jainja-tilck-20240912](https://github.com/user-attachments/assets/a89fb657-0c5d-42e7-bb00-2c56e1a2ddd8)

**Comment by vvaltchev** [comment]
Hello @glegris,
how does `jainja` itself work on Tilck if it's written in Java itself? Is there a way to statically link the OpenJDK runtime to an actual ELF binary?

I have no idea why it doesn't work. Tilck does not behave precisely like Linux does, even when it behaves as expected. Tilck behaves according to the interface implemented by Linux (POSIX + a few Linux-only special things), for the limited set of syscalls and syscall parameters it supports.

Many of the interesting applications I that ported on Tilck required quite some debugging and extra work in the kernel to work properly. If you statically linked somehow the Java runtime into an ELF binary, that's a ton of code and many things could have gone wrong. The thing might not work because of the following reasons:

- the limited syscall support or limited parameter support in Tilck, leading to unhandled errors in the user app
- a bug in Tilck's kernel
- a bug in the application
- POSIX but not "strictly Linux" behavior in Tilck while the app has unfair expectations based on Linux's implementation and not interface. 
  For example, a `read()` of a big buffer might read the whole buffer at once on Linux, while on Tilck the buffer size is likely smaller. Therefore, `read()` will return the number of bytes read and the caller has to continue to call `read()` in a loop. Sometimes C code assumes that a small-enough buffer will be read at once and don't code the loop. Another example is that on Tilck syscalls can fail with EINTR due to a signal, while on Linux that does not happen. That is POSIX behavior that apps should expect, but Linux-specific code might not do that.

Note that there might be multiple problems each in one of those categories. A starting point to debug that would be use Tilck's syscall tracing. See https://github.com/vvaltchev/tilck/blob/master/docs/debugging.md

Vlad

**Reply by glegris** [reply]
@vvaltchev, thanks for your reply.

The Jainja JVM is designed to be transpiled from Java to C. It can then be compiled using the Tilck toolchain. Only a subset of Java libraries is currently supported (~Java 5).

I will try to isolate the problem using debugging tools.
